src/main/java/org/galatea/starter/object/StockDay.java

```diff
@@ -0,0 +1,142 @@
+package org.galatea.starter.object;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.io.Serializable;
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.IdClass;
+import javax.persistence.Table;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import org.galatea.starter.service.object.AvResponse;
+import org.galatea.starter.utils.json.serialization.DateDeserializer;
+import org.galatea.starter.utils.json.serialization.DateSerializer;
+import org.galatea.starter.utils.json.serialization.MoneyDeserializer;
+
+@Data
+@Entity
+@Builder
+@Table(name = "price")
+@IdClass(StockSymbolId.class)
+@RequiredArgsConstructor
+@AllArgsConstructor
+public class StockDay implements Serializable, Comparable<StockDay> {
+
+  @Id
+  @JsonIgnore
+  private int stockId;
+  @JsonDeserialize(using = MoneyDeserializer.class)
+  @Column(name = "OpenPrice")
+  @JsonAlias(value = "1. open")
+  private BigDecimal open;
+  @JsonDeserialize(using = MoneyDeserializer.class)
+  @JsonAlias(value = "2. high")
+  @Column(name = "HighPrice")
+  private BigDecimal high;
+  @JsonDeserialize(using = MoneyDeserializer.class)
+  @JsonAlias(value = "3. low")
+  @Column(name = "LowPrice")
+  private BigDecimal low;
+  @JsonDeserialize(using = MoneyDeserializer.class)
+  @JsonAlias(value = "4. close")
+  @Column(name = "ClosePrice")
+  private BigDecimal close;
+  @JsonAlias(value = "5. volume")
+  private long volume;
+  @Id
+  @JsonSerialize(using = DateSerializer.class)
+  @JsonDeserialize(using = DateDeserializer.class)
+  @JsonProperty(value = "trade_date")
+  private OffsetDateTime tradeDate;
+  @JsonIgnore
+  private LocalTime updateTime;
+
+  /**
+   * updates all of the stock symbols in the supplied list with data from the given update map.<br>
+   *   (if a matching date is found)
+   * @param sourceList the list you're updating
+   * @param stockId the id the stock in the database
+   * @param response your map of new data with which to update
+   * @return
+   */
+  public static List<StockDay> updateAll(final List<StockDay> sourceList,

```

Comment:
✔ Originally, this was done because hibernate would see two objects with the same composite id, and throw an exception rather than doing an update during a saveOrUpdate (this would happen whenever a date had data that came from both the database cache and the alphavantage api). I'm unsure if this is still an issue with JpaRepository.

❔ This also updates each entry from alphavantage to have the trade date as a field rather than an object name before writing it to the database. I've updated the javadoc to be more clear about this. The only other way I can think to embed the key as a value would be to do it at acquisition time from the api with a custom deserializer for the entire map, and that sounds like a lot more work.

Reply:
hmmm... that's a pretty lame limitation of JPA lol.  If that's the case though, you could do a group delete and group save, right?  And for part 2 of your comment: if you use my suggestion from another comment to separate the StockDay class into two classes, you could add the date logic to the step that translates from the response model object to the domain object.

*important note though:*  You should change the name of the "tradeDate" field to something else.  In a lot of the work we do the "trade date" means the date a single deal is agreed upon. Whereas the date here is a representation of the date the stock was trading at this price.  So just calling it "date" or maybe "priceDate" would be better.

Reply:
✔looks like JpaRepository didn't have this same restriction - was able to eliminate updateAll by separating out StockDay into two classes per your other comment. (also, renamed trade_date to event_date)